src/v/kafka/server/connection_context.cc

```diff
@@ -186,32 +186,71 @@ bool connection_context::is_finished_parsing() const {
     return conn->input().eof() || _server.abort_requested();
 }
 
+connection_context::delay_t connection_context::calculate_throttle_delays(
+  const request_header& hdr, const size_t request_size) {
+    using clock = quota_manager::clock;
+    static_assert(std::is_same_v<clock, delay_t::clock>);
+
+    // Throttle on client based quotas
+    const auto client_quota_delay = _server.quota_mgr().record_tp_and_throttle(
+      hdr.client_id, request_size);
+
+    // Throttle on shard wide quotas
+    const auto now = clock::now();
```

Comment:
Just a thought, the API for existing throttle byte quotas is `record_tp_and_throttle` so the logic to calculate amount of quota is contained within the `quota_manager` itself.

Would it make sense for some of the logic here after `record_request_tp` to be moved into `quota_manager` as well? IMO keeping as much throttle related stuff in there is a win.

Reply:
I think about taking the entire `calculate_throttle_delays()` piece into `quota_manager`. I'm going to rebase onto the merged #7658 and see what I can do there.

Reply:
done